src/Concordium/Client/GRPC2.hs

```diff
@@ -0,0 +1,2658 @@
+        _nusnLevel1Keys <- fromProto $ nums ^. ProtoFields.level1Keys
+        _nusnLevel2Keys <- fromProto $ nums ^. ProtoFields.level2Keys
+        _nusnProtocol <- fromProto $ nums ^. ProtoFields.protocol
+        _nusnElectionDifficulty <- fromProto $ nums ^. ProtoFields.electionDifficulty
+        _nusnEuroPerEnergy <- fromProto $ nums ^. ProtoFields.euroPerEnergy
+        _nusnMicroCCDPerEuro <- fromProto $ nums ^. ProtoFields.microCcdPerEuro
+        _nusnFoundationAccount <- fromProto $ nums ^. ProtoFields.foundationAccount
+        _nusnMintDistribution <- fromProto $ nums ^. ProtoFields.mintDistribution
+        _nusnTransactionFeeDistribution <- fromProto $ nums ^. ProtoFields.transactionFeeDistribution
+        _nusnGASRewards <- fromProto $ nums ^. ProtoFields.gasRewards
+        _nusnPoolParameters <- fromProto $ nums ^. ProtoFields.poolParameters
+        _nusnAddAnonymityRevoker <- fromProto $ nums ^. ProtoFields.addAnonymityRevoker
+        _nusnAddIdentityProvider <- fromProto $ nums ^. ProtoFields.addIdentityProvider
+        _nusnCooldownParameters <- fromProto $ nums ^. ProtoFields.cooldownParameters
+        _nusnTimeParameters <- fromProto $ nums ^. ProtoFields.timeParameters
+        return QueryTypes.NextUpdateSequenceNumbers{..}
+
+instance FromProto Proto.IpAddress where
+    type Output Proto.IpAddress = IpAddress
+    fromProto peer = return . IpAddress $ peer ^. ProtoFields.value
+
+instance FromProto Proto.Port where
+    type Output Proto.Port = IpPort
+    fromProto port = do
+        -- Ensure that the value fits into a Word16.
+        let p = port ^. ProtoFields.value
+        let fits = fromIntegral p <= (maxBound :: Word16)
+        unless fits Nothing
+        return . IpPort $ fromIntegral p
+
+instance FromProto Proto.BannedPeer where
+    type Output Proto.BannedPeer = Peer
+    fromProto peer = fromProto $ peer ^. ProtoFields.ipAddress
+
+instance FromProto Proto.BannedPeers where
+    type Output Proto.BannedPeers = [Peer]
+    fromProto peers = mapM fromProto $ peers ^. ProtoFields.peers
+
+instance FromProto Proto.InstanceStateValueAtKey where
+    type Output Proto.InstanceStateValueAtKey = ByteString
+    fromProto value = return $ value ^. ProtoFields.value
+
+instance FromProto Proto.Signature where
+    type Output Proto.Signature = Signature
+    fromProto value = return . Signature $ BSS.toShort $ value ^. ProtoFields.value
+
+instance FromProto Proto.SignatureMap where
+    type Output Proto.SignatureMap = Updates.UpdateInstructionSignatures
+    fromProto uis = do
+        signatures <-
+            fmap Map.fromAscList
+                . mapM deMk
+                . Map.toAscList
+                $ uis ^. ProtoFields.signatures
+        return Updates.UpdateInstructionSignatures{..}
+      where
+        deMk (k, s) = do
+            -- Ensure that the value fits into a Word16.
+            if fromIntegral k <= (maxBound :: Word16)
```

Comment:
This check will always succeed since you are first converting a value to Word16, and then checking if the value is less than maxBound :: Word16. Which is always true.

You need to
```
k <= fromIntegral (maxBound :: Word16)
```

instead.

Reply:
Right! Nice catch.